### PR TITLE
feat(metro-config): add Tailwind × PostCSS support to Metro web

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

Tailwind is very popular and important to the web ecosystem, PostCSS is the easiest and most powerful way to configure it.

# How

Metro by default doesn't work with Tailwind and PostCSS because the PostCSS file is mutating an unchanging `global.css` file based on a set of inputs that Metro isn't aware of. This PR patches Metro so any file that is run through PostCSS is ignored in the Metro cache and processed on every request. We may need to optimize this in the future, but even with hundreds of files it wouldn't be the slowest part of Metro (`metro-file-map`, `metro-resolver`, Babel, etc.).

I've also updated the docs to reflect that configuration is now the exact same as any other web framework!

# Test Plan

I've added an E2E test, but it doesn't fully cover the HMR case. We don't have any current system for this, so I tested it manually by running `E2E_ROUTER_SRC=tailwind-postcss yarn expo start` in `apps/router-e2e` then modifying `apps/router-e2e/__e2e__/tailwind-postcss/app/index.tsx` and seeing the updates live.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 1]
```diff
@@ -219,26 +220,13 @@ Then, ensure [CSS is setup](#css) in the **metro.config.js** file.
 
 ### Tailwind
 
-> **info** Tailwind does not support native platforms, this is web-only.
+> **info** Standard Tailwind does not support native platforms, this is web-only. Use a compatability library like [Nativewind](https://www.nativewind.dev/) for universal support.
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:223]:
```suggestion
> **info** Standard Tailwind does not support native platforms. This is web-only. For universal support, use a compatible library like [Nativewind](https://www.nativewind.dev/).
```

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 2]
```diff
@@ -219,26 +220,13 @@ Then, ensure [CSS is setup](#css) in the **metro.config.js** file.
...
 
-[Tailwind](https://tailwindcss.com/) can be used with Metro for web. However, due to the advanced caching system in Metro, the setup is a little different from the default Tailwind setup. The following files are modified:
+[Tailwind](https://tailwindcss.com/) can be used with Metro for web. The following files are modified:
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:225]:
```suggestion
[Tailwind](https://tailwindcss.com/) can be used with Metro for web. The following files are required to be modified to set it up:
```

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 3]
```diff
@@ -219,26 +220,13 @@ Then, ensure [CSS is setup](#css) in the **metro.config.js** file.
...
-  ]}
-/>
-
-To setup, install the `tailwindcss` package in your project:
-
-<Terminal cmd={['$ yarn add -D tailwindcss@^3.3.1']} />
+<FileTree files={['app.json', 'package.json', 'global.css', 'tailwind.config.js', 'index.js']} />
 
-In your **app.json** ensure the project is using Metro for web:
+First, ensure you're using Metro for web in your **app.json**:
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:229]:
```suggestion
First, ensure you are using Metro for web in your **app.json**:
```

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 4]
```diff
@@ -250,16 +238,34 @@ In your **app.json** ensure the project is using Metro for web:
 }
 ```
 
-Create **global.css** in your project:
+Then simply configure Tailwind according to the [Tailwind PostCSS docs](https://tailwindcss.com/docs/installation/using-postcss).
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:241]:
```suggestion
Then, configure Tailwind according to the [Tailwind PostCSS docs](https://tailwindcss.com/docs/installation/using-postcss).
```

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 5]
```diff
@@ -250,16 +238,34 @@ In your **app.json** ensure the project is using Metro for web:
...
 
-```css global.css
-/* This file adds the requisite utility classes for Tailwind to work. */
-@tailwind base;
-@tailwind components;
-@tailwind utilities;
+<Step label="1">
+
+Install `tailwindcss` and its peer dependencies, and create your **tailwind.config.js** file.
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:245]:
```suggestion
Install `tailwindcss` and its peer dependencies, and create a **tailwind.config.js** file:
```

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 6]
```diff
@@ -250,16 +238,34 @@ In your **app.json** ensure the project is using Metro for web:
...
+
+<Terminal cmd={['$ yarn add --dev tailwindcss postcss autoprefixer', '$ npx tailwindcss init']} />
+
+</Step>
+
+<Step label="2">
+
+Add `tailwindcss` and `autoprefixer` to your **postcss.config.js** file.
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:253]:
```suggestion
Add `tailwindcss` and `autoprefixer` to your **postcss.config.js** file:
```

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 7]
```diff
@@ -250,16 +238,34 @@ In your **app.json** ensure the project is using Metro for web:
...
+  },
+};
 ```
 
-Create **tailwind.config.js** in your project:
+</Step>
+
+<Step label="3">
+
+Add the paths to all of your template files in your **tailwind.config.js** file.
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:268]:
```suggestion
Add the paths to all of your template files in your **tailwind.config.js** file:
```

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 8]
```diff
@@ -277,44 +283,39 @@ module.exports = {
...
-  // If this require.context is not inside the root directory (next to the package.json) then adjust this file path
-  // to resolve correctly.
-  './node_modules/.cache/expo/tailwind'
-);
-if (ctx.keys().length) ctx(ctx.keys()[0]);
+```js index.js
+import './global.css';
 ```
 
+> Expo Router users can use `./index.js` or `./app/_layout.js`, they both work the same.
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:307]:
```suggestion
> When using Expo Router, you can use **./index.js** or **./app/_layout.js**. They both work the same.
```

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 9]
```diff
@@ -277,44 +283,39 @@ module.exports = {
...
+
+</Step>
+
+<Step label="5">
+
+Start your project, you may want to clear the transform cache for good measure:
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:313]:
```suggestion
Start your project by clearing the transform cache for good measure:
```

File: docs/pages/versions/unversioned/config/metro.mdx [hunk 10]
```diff
@@ -343,6 +344,33 @@ export default function Page() {
 }
 ```
 
+#### Tailwind Troubleshooting
+
+If you have a custom `config.cacheStores` in your **metro.config.js**, you'll need to extend the Expo superclass of `FileStore`:
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/config/metro.mdx:349]:
```suggestion
If you have a custom `config.cacheStores` in your **metro.config.js**, you need to extend the Expo superclass of `FileStore`:
```

**Review by amandeepmittal — APPROVED** [review]

**Comment by marklawlor** [comment]
I'm wondering we should create a new docs page about `CSS` support. I feel its a bit mixed with up with the Metro docs. E.g If I was troubleshooting CSS hot reloading, this isn't the page I would expect the information. I think this is important, because nobody is going to associate a custom `FileStore` with breaking CSS hot-reload

**Pull request merged**